File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : './input.txt').toString().trim().split('\n');

const [N, M] = input[0].split(' ').map(Number);
const trees = input[1].split(' ').map(Number);
let mx = 0; 

const check = (h) => {
  return trees.reduce((acc, cur) => acc + Math.max(cur - h, 0), 0) >= M;
}

const answer = () => { 
  let ret = 0; 
  trees.forEach(tree => mx = Math.max(mx, tree));

  let lo = 0, hi = mx;
  
  
  while(lo <= hi){
    const mid = Math.floor((lo + hi) / 2);
    
    if(check(mid)){
        ret = Math.max(ret, mid);
        lo = mid + 1;
    }else{
        hi = mid - 1;
    }
  }

  console.log(ret);
}

answer();

File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

ll n, m, mx = 0, ret = 0, a[1000004];

bool bs(ll h) {
  ll ret = 0;
  for (int i = 0; i < n; i++) {
    if (a[i] - h > 0) ret += a[i] - h;
  }

  return ret >= m;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  cin >> n >> m;
  for (int i = 0; i < n; i++) {
    cin >> a[i];
    mx = max(mx, a[i]);
  }

  ll l = 0, r = mx;

  while (l <= r) {
    ll mid = (l + r) / 2;
    // cout << mid << ' ';

    if (bs(mid)) {
      l = mid + 1;
      ret = mid;
    } else {
      r = mid - 1;
    }
  }

  cout << ret;
  return 0;
}